### PR TITLE
[e2e] User creation + auth

### PR DESCRIPTION
**Pull request by Teddarific**
Title

### CODE REVIEW

File: server/app/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import { tokenForUser } from '../services/passport';
+
+export const createUser = async (req, res, next) => {
+  const { email, password, firstName, lastName } = req.body;
+
+  const errorMessages = {};
+
+  const emailTaken = await UserModel.findOne({ email });
+  if (emailTaken) {
+    errorMessages.email = 'This email is already taken.';
```
Comment (alexdanilowicz) [inline server/app/controllers/userController.js:12]: It's pretty cool to see how we write UserControllers has evolved over the years. Like comparing this to how it was done in a CS 52 lab... nice man.

File: server/app/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  });
+
+  try {
+    const savedUserModel = await newUser.save();
+    res.status(200).json({
+      token: tokenForUser(savedUserModel.id),
+      user: savedUserModel,
+    });
+  } catch (e) {
+    console.error(e);
```
Comment (alexdanilowicz) [inline server/app/controllers/userController.js:48]: Would the idea here be to formally log this error?

Obviously logging is not something we care about now... but just thinking long term and how we wanna do this "right"

File: server/app/models/userModel.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  {
+    JSON: {
+      virtuals: true,
+    },
+  }
+);
+
+UserSchema.method('toJSON', function () {
+  const obj = this.toObject();
+  delete obj.password;
```
Comment (alexdanilowicz) [inline server/app/models/userModel.js:21]: Is the idea here you're just deleting the $819y58291y5981y51892y5981y5 garbage whenever we ask for a json version of the obj?

File: server/app/services/validators.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+const EMAIL_REGEX = /[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,3}/;
```
Comment (alexdanilowicz) [inline server/app/services/validators.js:1]: clean. nice

File: web/package.json [hunk 5]
```diff
@@ -7,8 +7,14 @@
...
     "dependencies": {
+        "@material-ui/core": "^4.9.11",
+        "@material-ui/icons": "^4.9.1",
+        "axios": "^0.19.2",
         "react": "^16.13.1",
-        "react-dom": "^16.13.1"
+        "react-dom": "^16.13.1",
+        "react-redux": "^7.2.0",
+        "react-router-dom": "^5.1.2",
+        "redux": "^4.0.5"
```
Comment (alexdanilowicz) [inline web/package.json:17]: 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀

File: web/src/components/Core/Page.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import * as React from 'react'
+
+import CircularProgress from '@material-ui/core/CircularProgress'
+import Fade from '@material-ui/core/Fade'
+import { makeStyles } from '@material-ui/core/styles'
+
+const useStyles = makeStyles((theme) => ({
```
Comment (alexdanilowicz) [inline web/src/components/Core/Page.js:7]: are we making a pact to do no separate .scss files this time around?

File: web/src/components/Core/Page.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  root: {
+    backgroundColor: theme.palette.background.default,
+    minHeight: '100vh',
```
Comment (alexdanilowicz) [inline web/src/components/Core/Page.js:10]: im pretty sure if we are using vh and vw we need to be sure to add this to the html

`<meta name="viewport" content="width=device-width, initial-scale=1">`

I can take that as an action item

File: web/src/components/User/SignIn/SignInForm.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  field: {
+    margin: '8px 0',
+  },
+  submitButton: {
+    marginTop: '24px',
+    width: '200px',
+  },
+})
+
+function SignInForm(props) {
```
Comment (alexdanilowicz) [inline web/src/components/User/SignIn/SignInForm.js:39]: are we gonna have a rule that is H00Ks only aka no class based components

File: web/src/components/User/SignUp/SignUpForm.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
+import * as React from 'react'
+
+import Button from '@material-ui/core/Button'
+import CircularProgress from '@material-ui/core/CircularProgress'
+import FormControl from '@material-ui/core/FormControl'
+import FormHelperText from '@material-ui/core/FormHelperText'
+import IconButton from '@material-ui/core/IconButton'
```
Comment (alexdanilowicz) [inline web/src/components/User/SignUp/SignUpForm.js:7]: Nice job defining all of these out to keep the bundle small!

File: web/src/theme.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import { createMuiTheme, responsiveFontSizes } from '@material-ui/core/styles'
+
+// TODO: These are the wrong Inboard colors
```
Comment (alexdanilowicz) [inline web/src/theme.js:3]: ALL HAIL VLOGR!

**Review by alexdanilowicz — APPROVED** [review]

File: web/src/theme.js (hunk 10 above)
Comment (alexdanilowicz) [reply web/src/theme.js:3]: fyi i updated the colors on my branch so dont worry about it

File: server/app/controllers/userController.js (hunk 2 above)
Comment (Teddarific) [reply server/app/controllers/userController.js:48]: Sure, we can do really do whatever we want with this - no idea what we want long term.

File: server/app/models/userModel.js (hunk 3 above)
Comment (Teddarific) [reply server/app/models/userModel.js:21]: Yeah exactly

File: web/src/components/Core/Page.js (hunk 6 above)
Comment (Teddarific) [reply web/src/components/Core/Page.js:7]: yup

File: web/src/components/Core/Page.js (hunk 7 above)
Comment (Teddarific) [reply web/src/components/Core/Page.js:10]: Nice, TIL

File: web/src/components/User/SignIn/SignInForm.js (hunk 8 above)
Comment (Teddarific) [reply web/src/components/User/SignIn/SignInForm.js:39]: Well there's a time and place to use each, hooks don't completely replace class based components. I'd start everything off as a hook-based component, and only switch to class if you find that you need it

Comment (alexdanilowicz) [reply web/src/components/User/SignIn/SignInForm.js:39]: idk i had a chat with a buddy and he was talking about how like hooks are the future. Do you have an example were you would need a class? 

I guess other than if you need weird uncommon lifecycle methods? 

https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both 

**Do Hooks cover all use cases for classes?**
`Our goal is for Hooks to cover all use cases for classes as soon as possible. There are no Hook equivalents to the uncommon getSnapshotBeforeUpdate, getDerivedStateFromError and componentDidCatch lifecycles yet, but we plan to add them soon.`

Comment (alexdanilowicz) [reply web/src/components/User/SignIn/SignInForm.js:39]: https://stackoverflow.com/questions/54912470/will-react-hooks-replace-classes

File: server/app/controllers/userController.js (hunk 2 above)
Comment (alexdanilowicz) [reply server/app/controllers/userController.js:48]: re: logging, a coworker should be this blog today written by founder of heroku, thought you mind find interesting: https://12factor.net/logs

If you haven't seen https://12factor.net/ it's pretty nice at a high description of best practices/scale (but of course nothing is dogma)

File: web/src/components/User/SignIn/SignInForm.js (hunk 8 above)
Comment (alexdanilowicz) [reply web/src/components/User/SignIn/SignInForm.js:39]: @Teddarific bump legitly curious if you had any thoughts on when you'd use a class-based component over a hook unless u were doing some really funky, uncommon thing?

**Pull request merged**